File: tasks-core-jvm/src/main/java/org/funfix/tasks/Trampoline.java
package org.funfix.tasks;

import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.util.LinkedList;
import java.util.concurrent.Executor;

@NullMarked
final class Trampoline {
    private Trampoline() {}

    private static final ThreadLocal<@Nullable LinkedList<Runnable>> queue =
            new ThreadLocal<>();

    private static void eventLoop() {
        while (true) {
            final var current = queue.get();
            if (current == null) {
                return;
            }
            final var next = current.pollFirst();
            if (next == null) {
                return;
            }
            try {
                next.run();
            } catch (final Exception e) {
                UncaughtExceptionHandler.logOrRethrowException(e);
            }
        }
    }

    public static final Executor INSTANCE =
            command -> {
                var current = queue.get();
                if (current == null) {
                    current = new LinkedList<>();
                    current.add(command);
                    queue.set(current);
                    try {
                        eventLoop();
                    } finally {
                        queue.remove();
                    }
                } else {
                    current.add(command);
                }
            };

    public static void execute(final Runnable command) {
        INSTANCE.execute(command);
    }
}

File: tasks-core/src/jvmMain/kotlin/org/funfix/tasks/Trampoline.kt
package org.funfix.tasks

import java.util.LinkedList
import java.util.concurrent.Executor
import kotlin.concurrent.getOrSet

/**
 * [Executor] implementation that uses a thread-local trampoline to
 * execute tasks in a stack-safe way.
 *
 * INTERNAL API.
 */
internal object Trampoline {
    private val queue = ThreadLocal<LinkedList<Runnable>>()

    private fun eventLoop() {
        while (true) {
            val current = queue.get()
            if (current == null || current.isEmpty()) {
                return
            }
            val next = current.pollFirst()
            try {
                next?.run()
            } catch (e: Exception) {
                UncaughtExceptionHandler.logException(e)
            }
        }
    }

    @JvmField
    val EXECUTOR: Executor = Executor { command ->
        val current = queue.getOrSet { LinkedList() }
        current.add(command)
        if (current.size == 1) {
            try {
                eventLoop()
            } finally {
                queue.remove()
            }
        }
    }

    @JvmStatic
    fun execute(command: Runnable) {
        EXECUTOR.execute(command)
    }
}
